design-proposals/klip-4-custom-types.md

```diff
@@ -0,0 +1,186 @@
+  # KLIP-4: Custom Type Registry
+
+**Author**: agavra | 
+**Release Target**: 5.4 | 
+**Status**: In Discussion | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/2894
+
+**tl;dr:** *Introduce a feature that makes custom types easier to work with in KSQL by aliasing
+complex type declarations.*
+
+## Motivation and background
+
+Kafka often contains very complex nested data structures and it is common to see schemas nested with
+other schemas. It is desirable to be able to define this custom schema in KSQL once and reference it
+in all future statements. For example, imagine the following schema declarations:
+
+```sql
+CREATE TYPE ADDRESS AS STRUCT<number INTEGER, street VARCHAR, city VARCHAR>;
+CREATE TYPE PERSON AS STRUCT<firstname VARCHAR, lastname VARCHAR, address ADDRESS>;
+CREATE TYPE COMPANY AS STRUCT<name VARCHAR, headquarters ADDRESS>;
+```
+
+With this feature organizations can leverage existing schemas in their KSQL queries and share 
+data with other parts of their organizations.
+
+## Scope
+
+* Syntax to register and unregister custom type aliases
+* Ability to include custom types at compile-time (e.g. in the JAR) for use in UDF declarations
+* Ability to list all custom types
+
+Schema Registry integration is not considered within scope.
+
+## Value/Return
+
+Real data is complex, nested and referential. This feature will help us dramatically improve the
+user interaction with complicated schemas. 
+
+## Public APIS
+
+### CREATE TYPE
+
+```sql
+CREATE TYPE <type_name> AS <type>;
+```
+
+The `CREATE TYPE` syntax will allow KSQL users to register a type alias directly in SQL (either 
+interactive or headless modes). Any types registered using this command can be leveraged in any
+future statement. 
+
+Any attempts to register the same type twice without a corresponding `DROP TYPE` will fail.
+
+### DROP TYPE 
+
+```sql
+DROP TYPE <type_name>;
+```
+
+The `DROP TYPE` syntax will allow KSQL users to remove a type alias from KSQL. This statement
+will fail if the type is being used in any active query or UDF.
```

Comment:
I think this goes back to whether we want soft or hard references. I think the ROI on soft references is lower than that of hard references (namely, we solve 99% of the problems with hard references) so the proposal in this KLIP is to use hard references.

If we use hard references, then it's just a convenience - in fact we should be able to disregard compatibility ("is in use") checks altogether.

Reply:
IN addition to queries and UDFs we should also make sure the type is not used in DDL statements if we are dropping it.

Reply:
By the way, @agavra can you elaborate on what you mean by hard and soft references?

Reply:
I removed this requirement for now since it's not really necessary with hard references. As andy pointed out in his comment https://github.com/confluentinc/ksql/pull/2894#issuecomment-503498109 we can extend it to emulate soft references in the future